### PR TITLE
Cegis unif enumerator manager

### PR DESCRIPTION
**Pull request by ajreynol**
This is in preparation for a CegisUnif approach which will try to find minimally diverse models for the set of evaluation points.

### CODE REVIEW

File: src/theory/quantifiers/sygus/cegis_unif.h [hunk 1]
```diff
@@ -152,6 +152,95 @@ class CegisUnif : public SygusModule
 
 }; /* class CegisUnif */
 
+/** Cegis Unif Enumeration Manager
+ *
+ * This function enforces a decision heuristic that limits the number of
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/cegis_unif.h]: function --> class

File: src/theory/quantifiers/sygus/cegis_unif.h [hunk 2]
```diff
@@ -152,6 +152,95 @@ class CegisUnif : public SygusModule
...
+   public:
+    TypeInfo() {}
+    /** candidates for this type */
+    std::vector<Node> d_candidates;
+    /** the set of enumerators we have allocated for this candidate */
+    std::vector<Node> d_enums;
+    /** the set of evaluation points of this type */
+    std::vector<Node> d_eval_points;
+  };
+  std::map<TypeNode, TypeInfo> d_ce_info;
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/cegis_unif.h]: doc

File: src/theory/quantifiers/sygus/cegis_unif.h [hunk 3]
```diff
@@ -152,6 +152,95 @@ class CegisUnif : public SygusModule
...
+  /** literals of the form G_uq_n for each n */
+  std::map<unsigned, Node> d_guq_lit;
+  /**
+   * The minimal n such that G_uq_n is asserted positively in the
+   * current SAT context.
+   */
+  context::CDO<unsigned> d_curr_guq_val;
+  /** increment the number of enumerators */
+  void incrementNumEnumerators();
+  /** get current cost fun literal */
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/cegis_unif.h]: describe cost fun here or elsewhere

File: src/theory/quantifiers/sygus/cegis_unif.h [hunk 4]
```diff
@@ -152,6 +152,95 @@ class CegisUnif : public SygusModule
...
+   * synthesis-by-unification approach for c.
+   */
+  void initialize(std::vector<Node>& cs);
+  /** register evaluation point for candidate
+   *
+   * This notifies this class that eis is a set of evaluation points for
+   * candidate c, where c should be a candidate that was passed to initialize
+   * in the vector cs.
+   */
+  void registerEvalPts(std::vector<Node>& eis, Node c);
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/cegis_unif.h]: clarify that adding evaluation points triggers the creation of the lemmas constraining their values according to the current cost function

File: src/theory/quantifiers/sygus/cegis_unif.cpp [hunk 5]
```diff
@@ -235,6 +235,133 @@ void CegisUnif::registerRefinementLemma(const std::vector<Node>& vars,
...
+}
+
+Node CegisUnifEnumManager::getOrMkLiteral(unsigned n)
+{
+  std::map<unsigned, Node>::iterator itc = d_guq_lit.find(n);
+  if (itc == d_guq_lit.end())
+  {
+    NodeManager* nm = NodeManager::currentNM();
+    // initialize it
+    Node new_lit = Rewriter::rewrite(nm->mkSkolem("G_cost", nm->booleanType()));
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/cegis_unif.cpp]: why do this rewrite?

File: src/theory/quantifiers/sygus/cegis_unif.cpp [hunk 6]
```diff
@@ -235,6 +235,133 @@ void CegisUnif::registerRefinementLemma(const std::vector<Node>& vars,
...
+    new_lit = d_qe->getValuation().ensureLiteral(new_lit);
+    AlwaysAssert(!new_lit.isNull());
+    d_qe->getOutputChannel().requirePhase(new_lit, true);
+    d_guq_lit[n] = new_lit;
+    // allocate an enumerator for each candidate
+    for (std::pair<const TypeNode, TypeInfo>& ci : d_ce_info)
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/cegis_unif.cpp]: I think it makes more sense for this cade to be ``incrementNumEnumerators`` than here

File: src/theory/quantifiers/sygus/cegis_unif.cpp (hunk 5 above)
Comment (ajreynol) [reply src/theory/quantifiers/sygus/cegis_unif.cpp]: the requirePhase function below requires that new_lit is in rewritten form.  This rewrite likely will never do anything, but it is for paranoid reasons.

File: src/theory/quantifiers/sygus/cegis_unif.cpp (hunk 6 above)
Comment (ajreynol) [reply src/theory/quantifiers/sygus/cegis_unif.cpp]: I agree, I've reorganized things a bit to make this clear.

**Pull request merged**